flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala

```diff
@@ -0,0 +1,349 @@
+  * against the best offers, leading to revised offers containing reserved resources with which to launch task(s).
+  */
+class LaunchCoordinator(
+    manager: ActorRef,
+    config: Configuration,
+    schedulerDriver: SchedulerDriver,
+    optimizerBuilder: TaskSchedulerBuilder
+  ) extends Actor with FSM[TaskState, GatherData] {
+
+  val LOG = Logger(getClass)
+
+  /**
+    * The task placement optimizer.
+    *
+    * The optimizer contains the following state:
+    *  - unused offers
+    *  - existing task placement (for fitness calculation involving task colocation)
+    */
+  private[mesos] val optimizer: TaskScheduler = {
+    optimizerBuilder
+      .withLeaseRejectAction(new Action1[VirtualMachineLease]() {
+        def call(lease: VirtualMachineLease) {
+          LOG.info(s"Declined offer ${lease.getId} from ${lease.hostname()} of ${lease.memoryMB()} MB, ${lease.cpuCores()} cpus.")
+          schedulerDriver.declineOffer(lease.getOffer.getId)
+        }
+      }).build
+  }
+
+  override def postStop(): Unit = {
+    optimizer.shutdown()
+    super.postStop()
+  }
+
+  /**
+    * Initial state
+    */
+  startWith(Suspended, GatherData(tasks = Nil, newLeases = Nil))
+
+  /**
+    * State: Suspended
+    *
+    * Wait for (re-)connection to Mesos.   No offers exist in this state, but outstanding tasks might.
+    */
+  when(Suspended) {
+    case Event(msg: Connected, data: GatherData) =>
+      if(data.tasks.nonEmpty) goto(GatheringOffers)
+      else goto(Idle)
+  }
+
+  /**
+    * State: Idle
+    *
+    * Wait for a task request to arrive, then transition into gathering offers.
+    */
+  onTransition {
+    case _ -> Idle => assert(nextStateData.tasks.isEmpty)
+  }
+
+  when(Idle) {
+    case Event(msg: Disconnected, data: GatherData) =>
+      goto(Suspended)
+
+    case Event(offers: ResourceOffers, data: GatherData) =>
+      // decline any offers that come in
+      schedulerDriver.suppressOffers()
+      for(offer <- offers.offers().asScala) { schedulerDriver.declineOffer(offer.getId) }
+      stay()
+
+    case Event(msg: Launch, data: GatherData) =>
+      goto(GatheringOffers) using data.copy(tasks = data.tasks ++ msg.tasks.asScala)
+  }
+
+  /**
+    * Transition logic to control the flow of offers.
+    */
+  onTransition {
+    case _ -> GatheringOffers =>
+      LOG.info(s"Now gathering offers for at least ${nextStateData.tasks.length} task(s).")
+      schedulerDriver.reviveOffers()
+
+    case GatheringOffers -> _ =>
+      // decline any outstanding offers and suppress future offers
+      LOG.info(s"No longer gathering offers; all requests fulfilled.")
+
+      assert(nextStateData.newLeases.isEmpty)
+      schedulerDriver.suppressOffers()
+      optimizer.expireAllLeases()
+  }
+
+  /**
+    * State: GatheringOffers
+    *
+    * Wait for offers to accumulate for a fixed length of time or from specific slaves.
```

Comment:
Why do we wait for offers being accumulated? Why not starting tasks as soon as we get offers?